khmarbaise/multienv-maven-plugin
# Filtering

Most of the time resource files for different environments are largely the same. They only have minor differences. In those scenarios it is quite cumborsome to parse through the entire file to find the small changes. It makes sense to put those minor differences in a properties files and move the full files to a common location. 

For example take **logback.xml** file. The file is usually only differs in **server address**, **port** and a few other things. It makes sense to move those small differences into a properties file in each of the environment directories and move a single logback.xml file into a common location. So the directory structure looks something like this.

- src/
	- main/
	   	- environments/ 			
		   	- common/
			   	- logback.xml
			- dev/
				- envspec.properties
			- ci/
				- envspec.properties
			- qa/
				- envspec.properties

Now all wars will have **logback.xml** and the **logback.xml** is filtered using **envspec.properties** for each environment individually.

But **multienv-maven** currently doesn't support custom filtering for each environment. Infact the readme file mentions this under filtering if I am not wrong.

	Different files for different environments

Although **AbstractMultiEnvMojo** has a field called **filters**, it isn't being used. This change Makes use of that field and it also introduces another configuration item called **commonDir**.

With this there will be two steps to filtering:

1. All files in **src/main/environments** are filtered using the files listed in the filter tags in build section. This includes the common directory. 

2. Then **multienv-maven** filters the common files using the **key/value pairs** from the properties file declared in **filters** tag inside the configuration segment of multienv maven plugin. If **filters** are declared but **commonDir** is not declared an error message will be logged telling the user that the filters declared here are only used to filter files in common directory. It also tells the user if he has no intention of using a common directory he should declare filters in the build section of the pom.

NOTE: This includes the changes from pull request https://github.com/khmarbaise/multienv-maven-plugin/pull/23

Having a common directory keeps things clean and easy to understand.
